modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/IndexManagerTest.java

```diff
@@ -0,0 +1,228 @@
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.data.shapefile;
+
+import static org.geotools.data.shapefile.files.ShpFileType.QIX;
+import static org.geotools.data.shapefile.files.ShpFileType.SHP;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.geotools.data.shapefile.files.ShpFiles;
+import org.geotools.feature.NameImpl;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class IndexManagerTest extends TestCaseSupport {
+
+    private ShapefileDataStore mockDataStore;
+    private ShpFiles shpFiles;
+
+    @Before
+    public void setUp() throws IOException {
+        File statesShp = copyShapefiles(STATE_POP);
+        shpFiles = new ShpFiles(statesShp);
+
+        mockDataStore = mock(ShapefileDataStore.class);
+        when(mockDataStore.getTypeName()).thenReturn(new NameImpl("statepop"));
+    }
+
+    @After
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        shpFiles.dispose();
+    }
+
+    @Test
+    public void testCreateSpatialIndex() throws Exception {
+        deleteQixFile();
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+
+        assertFalse(shpFiles.exists(QIX));
+        final boolean force = false;
+        assertTrue(indexManager.createSpatialIndex(force));
+        assertTrue(shpFiles.exists(QIX));
+        assertFalse(indexManager.createSpatialIndex(force));
+    }
+
+    @Test
+    public void testCreateSpatialIndex_force() throws Exception {
+        deleteQixFile();
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+
+        final boolean force = true;
+        assertTrue(indexManager.createSpatialIndex(force));
+
+        File file = null;
+        try {
+            file = shpFiles.acquireReadFile(QIX, indexManager.writer);
+            assertNotNull(file);
+        } finally {
+            shpFiles.unlockRead(file, indexManager.writer);
+        }
+
+        final long currentLastModified = file.lastModified();
+
+        for (int i = 0;
```

Comment:
A bit worried about this bit, in the context of build servers and PR checks... which can be very unreliable in terms of time taken to do things. 1 second is a lot of time on a dev machine, but might not be enough on an overworked build server or small PR check VM. We have been bitten by this kind of assumption several times, with tests assuming quick execution randomly failing every now and then.

I suggest the following change: allow the loop to run for eventually a long time (e.g., 30 seconds?), but bail out of the loop as soon as the condition you're looking for (new file created). If you exit the loop without the file being re-created, then it's a failiure.
With this approach it's going to be fast (even faster) on a dev machine, but it's lenient towards a slow PR check that's being kicked out of the CPUs by other VMs (or by other tests running in parallel, even the build checks run with -T2).

Reply:
Done! https://github.com/geotools/geotools/pull/3527/files#diff-cca0c8a31b2bd3b8cf2a2473aae2301caf3f80def759a5d36a2d82b3cc4e319bR94-R103